internal/statestore/backfill_test.go

```diff
@@ -300,6 +315,165 @@ func TestDeleteBackfill(t *testing.T) {
 	require.Error(t, err)
 	require.Equal(t, codes.Unavailable.String(), status.Convert(err).Code().String())
 	require.Contains(t, status.Convert(err).Message(), "DeleteBackfill, id: 12345, failed to connect to redis:")
+
+}
+
+// TestAcknowledgeBackfillLifecycle test statestore functions - AcknowledgeBackfill, GetExpiredBackfillIDs
+// and deleteExpiredBackfillID
+func TestAcknowledgeBackfillLifecycle(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+	ctx := utilTesting.NewContext(t)
+
+	bf1 := "mockBackfillID"
+	bf2 := "mockBackfillID2"
+	err := service.CreateBackfill(ctx, &pb.Backfill{
+		Id:         bf1,
+		Generation: 1,
+	}, nil)
+	require.NoError(t, err)
+
+	err = service.CreateBackfill(ctx, &pb.Backfill{
+		Id:         bf2,
+		Generation: 1,
+	}, nil)
+	require.NoError(t, err)
+
+	bfIDs, err := service.GetExpiredBackfillIDs(ctx)
+	require.NoError(t, err)
+	require.Len(t, bfIDs, 0)
+	pendingReleaseTimeout := cfg.GetDuration("pendingReleaseTimeout")
+
+	// Sleep till Backfills
+	time.Sleep(pendingReleaseTimeout)
+
+	// This call also sets initial LastAcknowledge time
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.NoError(t, err)
+	require.Len(t, bfIDs, 2)
+	require.Contains(t, bfIDs, bf1)
+	require.Contains(t, bfIDs, bf2)
+
+	err = service.AcknowledgeBackfill(ctx, bf1)
+	require.NoError(t, err)
+	err = service.AcknowledgeBackfill(ctx, bf2)
+	require.NoError(t, err)
+
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.NoError(t, err)
+	require.Len(t, bfIDs, 0)
+
+	// Sleep until the pending release expired and verify we have all the backfills
+	time.Sleep(pendingReleaseTimeout)
+
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.NoError(t, err)
+	require.Len(t, bfIDs, 2)
+	require.Contains(t, bfIDs, bf1)
+	require.Contains(t, bfIDs, bf2)
+
+	// Acknowledge one Backfill it should be removed from GetExpired output
+	err = service.AcknowledgeBackfill(ctx, bf2)
+	require.NoError(t, err)
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.Len(t, bfIDs, 1)
+	require.NoError(t, err)
+	require.Equal(t, bf1, bfIDs[0])
+
+	err = service.DeleteBackfill(ctx, bfIDs[0])
+	require.NoError(t, err)
+
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.Len(t, bfIDs, 0)
+	require.NoError(t, err)
+}
+
+func TestAcknowledgeBackfill(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+
+	startTime := time.Now()
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+	ctx := utilTesting.NewContext(t)
+	bf1 := "mockBackfillID"
+	err := service.CreateBackfill(ctx, &pb.Backfill{
```

Comment:
here you need to create backfill directly in redis to test AcknowledgeBackfill function separately, because CreateBackfill uses AcknowledgeBackfill under the hood.

Reply:
Not sure if backfill should be created here at all, it is not used anyway.
AcknowledgeBackfill accepts an id as a string.